File: src/main/scala/kim/dencka/easy/collection/ClimbingStairs.scala
package kim.dencka.easy.collection

object ClimbingStairs extends App {

  def climbStairs(n: Int): Int = {
    @annotation.tailrec
    def run(cur: Int, prev: Int, step: Int): Int = n match {
      case 1 => 1
      case 2 => 2
      case 3 => 3
      case _ if step != n => run(cur + prev, cur, step + 1)
      case _ => cur
    }
    run(3, 2, 3)
  }

  println(climbStairs(4))

}

File: src/main/scala/kim/dencka/java/easy/ClimbingStairs.java
package kim.dencka.java.easy;

class ClimbingStairs {
    public static void main(String[] args) {
        System.out.println(climbStairs(1));
        System.out.println(climbStairs(2));
        System.out.println(climbStairs(3));
        System.out.println(climbStairs(4));
        System.out.println(climbStairs(5));
    }

    public static int climbStairs(int n) {
        if (n == 1) return 1;
        if (n == 2) return 2;
        int i = 1, j = 2, step = 2;

        int res = 0;
        while (step != n) {
            res = i + j;
            i = j;
            j = res;
            step++;
        }

        return res;
    }
}
